packages/experimental/Button/src/Button.tsx

```diff
@@ -1,41 +1,55 @@
 /** @jsx withSlots */
 import * as React from 'react';
 import { Image, View } from 'react-native';
-import { ButtonProps, buttonName, ButtonType } from './Button.types';
+import { buttonName, ButtonType, ButtonTypeMac, ButtonMacProps, NativeButtonProps, ButtonProps } from './Button.types';
 import { Text } from '@fluentui-react-native/experimental-text';
 import { stylingSettings } from './Button.styling';
 import { filterImageProps } from '@fluentui-react-native/adapters';
 import { compose, mergeProps, withSlots, UseSlots } from '@fluentui-react-native/framework';
 import { useButton } from './useButton';
+import { ensureNativeComponent } from '@fluentui-react-native/component-cache';
+import { Platform } from 'react-native';
 
-export const Button = compose<ButtonType>({
-  displayName: buttonName,
-  ...stylingSettings,
-  slots: {
-    root: View,
-    icon: Image,
-    content: Text,
-  },
-  filters: {
-    icon: filterImageProps,
-  },
-  render: (userProps: ButtonProps, useSlots: UseSlots<ButtonType>) => {
-    const button = useButton(userProps);
-    // grab the styled slots
-    const Slots = useSlots(userProps, (layer) => button.state[layer] || userProps[layer]);
-    // now return the handler for finishing render
-    return (final: ButtonProps, ...children: React.ReactNode[]) => {
-      const { icon, content, ...mergedProps } = mergeProps(button.props, final);
+const NativeButton = ensureNativeComponent('MSFButtonView');
 
-      return (
-        <Slots.root {...mergedProps}>
-          {icon && <Slots.icon key="icon" source={{ uri: icon }} />}
-          {content && <Slots.content key="content">{content}</Slots.content>}
-          {children}
-        </Slots.root>
-      );
-    };
-  },
-});
+export const Button =
+  Platform.OS === 'macos'
+    ? compose<ButtonTypeMac>({
```

Comment:
Right now I don't think this approach is a great experience for customers. (See the forking required in the test code) Fundamentally we are making the macos button inexplicably have a different and less functional interface. There are issues with each:

- Different - the fact that the interfaces are different is an incredible pain for people authoring as they have to fork/re-author everything. Typescript also doesn't have good affordances for multi-platform building. As a result it is certain that TS builds are only validating for one platform, type errors will slip through the other ones.
- Less Functional - taking recomposition scenarios out of it for a second (as those are hard), this effectively only allows a few pre-canned button variants, ignores tokens entirely (though it will still have a customize function on it). This also means that if you author your code in win32 first (which many Office feature authors are doing), then try to port it over to mac, you'll have to come up with new implementations for any custom button usages.

I think there are really two viable approaches here. Having Button be a unified front-end that will transparently use native buttons internally when they are sufficient for what the customer wants, or make a well defined native button control that is explicit in what you are using and supporting.

I'd probably start by creating an explicit native button control. If you are using native button anyways you suddenly have a tie between native/JS code that didn't necessarily exist before. (If you service deliver a new bundle you have to validate the versioning as the native code won't update)

The explicit control allows thinking through what an appropriate interface looks like, thinking through questions like:

- is this a native button for all platforms? All apple platforms? Mobile? Mac only?
- This leads into what is the best interface for a native button
- What is our versioning story for matching our native bits to our bundle? (I've been diving into SDK stuff for the past few weeks so this ends up being a big deal)

We can then look at routing button to native when it is available but that might be a step 2.

Reply:
I created a separate component for native button that consists of the native implementations. Right now, it only supports apple platforms, but I think it has the potential to support other platforms as well. Since the control takes a dependency on FluentUI native, we need to update the native version in FURN whenever we make changes to the native button.